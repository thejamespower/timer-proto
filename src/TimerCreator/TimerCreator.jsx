import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import TimeField from 'react-simple-timefield'

class TimerCreator extends Component {
  constructor(props) {
    super(props)
    this.state = {
      time: '00:00',
    }

    this.handleTimeChange = this.handleTimeChange.bind(this)
    this.handleClick = this.handleClick.bind(this)
  }

  handleTimeChange(time) {
    this.setState({ time })
  }

  handleClick() {
    const { createdTimer } = this.props
    const { time } = this.state
    const duration = moment.duration(time)

    createdTimer({
      date: moment().add(duration),
      id: moment().format(),
    })
  }

  render() {
    const { time } = this.state

    return (
      <div>
        <TimeField value={time} onChange={this.handleTimeChange} showSeconds />
        <button
          type="submit"
          onClick={() => {
            this.handleClick()
          }}
        >
Create timer
        </button>
      </div>
    )
  }
}

TimerCreator.propTypes = {
  createdTimer: PropTypes.func.isRequired,
}

export default TimerCreator
