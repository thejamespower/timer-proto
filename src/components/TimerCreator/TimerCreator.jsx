import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';

import { TextField } from '@material-ui/core';
import CustomTimeField from '../TimeField';
import convertSecondsToDuration from '../../lib/convertSecondsToDuration';
import convertDurationToSeconds from '../../lib/convertDurationToSeconds';

const uuid = require('uuid/v4');

const zeroDuration = '00:00:00';

class TimerCreator extends Component {
  static propTypes = {
    createTimer: PropTypes.func.isRequired,
    createSubTimer: PropTypes.func.isRequired,
    name: PropTypes.string.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      duration: zeroDuration,
      subTimers: [],
    };

    this.changeDuration = this.changeDuration.bind(this);
    this.addTimerToSuperTimer = this.addTimerToSuperTimer.bind(this);
    this.createSubTimer = this.createSubTimer.bind(this);
    this.changeSubTimerDuration = this.changeSubTimerDuration.bind(this);
    this.changeSubTimerName = this.changeSubTimerName.bind(this);
  }

  changeDuration(duration) {
    this.setState({ duration });
  }

  changeSubTimerDuration(subTimer, duration) {
    const { subTimers } = this.state;
    this.setState({
      subTimers: subTimers.map(timer =>
        timer.id === subTimer.id
          ? {
              ...timer,
              duration,
            }
          : timer,
      ),
    });
  }

  changeSubTimerName(subTimer, name) {
    const { subTimers } = this.state;
    this.setState({
      subTimers: subTimers.map(timer =>
        timer.id === subTimer.id
          ? {
              ...timer,
              name,
            }
          : timer,
      ),
    });
  }

  addTimerToSuperTimer() {
    const { createTimer, name, createSubTimer } = this.props;
    const { duration, subTimers } = this.state;

    if (duration === zeroDuration) return;

    const parentId = uuid();

    createTimer({
      duration,
      id: parentId,
      name,
    });

    subTimers.map(subTimer =>
      createSubTimer({
        duration: subTimer.duration,
        id: uuid(),
        parentId,
        name: `${subTimer.name} ${name}`,
        offset: convertSecondsToDuration(
          convertDurationToSeconds(duration) -
            convertDurationToSeconds(subTimer.duration),
        ),
      }),
    );
  }

  createSubTimer() {
    const { subTimers } = this.state;

    this.setState({
      subTimers: [
        ...subTimers,
        {
          offset: zeroDuration,
          duration: zeroDuration,
          id: uuid(),
          name: '',
        },
      ],
    });
  }

  removeSubTimer(id) {
    const { subTimers } = this.state;

    this.setState({
      subTimers: subTimers.filter(timers => timers.id !== id),
    });
  }

  render() {
    const { duration, subTimers } = this.state;

    return (
      <div>
        <CustomTimeField value={duration} onChange={this.changeDuration} />

        {subTimers.map(subTimer => (
          <div key={subTimer.id}>
            <Button
              type="button"
              onClick={() => {
                this.removeSubTimer(subTimer.id);
              }}
              style={{ marginTop: '1rem' }}
            >
              X
            </Button>
            <TextField
              style={{ width: '100%' }}
              id="item-name"
              label="Name"
              value={subTimer.name}
              onChange={event =>
                this.changeSubTimerName(subTimer, event.target.value)
              }
              margin="normal"
            />
            <CustomTimeField
              value={subTimer.duration}
              onChange={value => this.changeSubTimerDuration(subTimer, value)}
            />
          </div>
        ))}
        <Button
          type="button"
          onClick={() => {
            this.createSubTimer();
          }}
          style={{ marginTop: '1rem' }}
        >
          Do something else to it?
        </Button>

        <Button
          type="submit"
          onClick={() => {
            this.addTimerToSuperTimer();
          }}
          style={{ marginTop: '1rem' }}
        >
          Create timer
        </Button>
      </div>
    );
  }
}

export default TimerCreator;
