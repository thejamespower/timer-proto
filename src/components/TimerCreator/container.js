import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'

import TimerCreator from './TimerCreator'
import { createdTimer } from '../../action-creators'

const mapStateToProps = null

const mapDispatchToProps = dispatch => bindActionCreators({
  createdTimer,
}, dispatch)

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TimerCreator)
