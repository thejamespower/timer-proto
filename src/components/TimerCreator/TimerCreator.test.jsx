import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Button from '@material-ui/core/Button'
import TimerCreator from './TimerCreator'
import CustomTimeField from '../TimeField'

describe('TimerCreator', () => {
  let props
  beforeEach(() => {
    props = {
      createTimer: () => null,
      name: 'test',
    }
  })

  describe('render', () => {
    it('renders', () => {
      const component = renderer.create(
        <TimerCreator {...props} />,
      )
      const tree = component.toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('behaviour', () => {
    it('handles CustomTimeField onChange', () => {
      const duration = '00:00:01'
      const wrapper = shallow(<TimerCreator {...props} />)

      wrapper.instance().handleTimeChange = jest.fn()
      wrapper.instance().setState({ duration })
      const customTimeField = wrapper.find(CustomTimeField)

      customTimeField.prop('onChange')(duration)

      expect(wrapper.instance().handleTimeChange).toHaveBeenCalledWith(duration)
    })

    it('handles handleTimeChange', () => {
      const duration = '00:00:01'
      const wrapper = shallow(<TimerCreator {...props} />)

      wrapper.setState({ duration }, () => {
        wrapper.instance().setState = jest.fn()
        wrapper.instance().handleTimeChange(duration)
        expect(wrapper.instance().setState).toHaveBeenCalledWith({ duration })
      })
    })

    it('handles Button onClick', () => {
      const wrapper = shallow(<TimerCreator {...props} />)
      const button = wrapper.find(Button)

      wrapper.instance().handleClick = jest.fn()
      button.prop('onClick')()

      expect(wrapper.instance().handleClick).toHaveBeenCalled()
    })

    it('handles handleClick with invalid duration', () => {
      props.createTimer = jest.fn()
      const wrapper = shallow(<TimerCreator {...props} />)

      wrapper.instance().handleClick()

      expect(props.createTimer).not.toHaveBeenCalled()
    })

    it('handles handleClick with valid duration', () => {
      const duration = '00:00:01'
      props.createTimer = jest.fn()
      const wrapper = shallow(<TimerCreator {...props} />)

      wrapper.setState({ duration }, () => {
        wrapper.instance().handleClick()

        expect(props.createTimer).toHaveBeenCalledWith({
          name: props.name,
          duration,
          id: expect.any(String),
        })
      })
    })
  })
})
