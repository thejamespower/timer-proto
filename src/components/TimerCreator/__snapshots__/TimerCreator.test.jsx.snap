// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimerCreator renders 1`] = `
<div>
  <TimeField
    onChange={[Function]}
    value="00:00:00"
  />
  <button
    className="MuiButtonBase-root-27 MuiButton-root-1 MuiButton-text-3 MuiButton-flat-6"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="submit"
  >
    <span
      className="MuiButton-label-2"
    >
      Create timer
    </span>
    <span
      className="MuiTouchRipple-root-30"
    />
  </button>
</div>
`;
